theme-ui/index.js

```diff
@@ -1,14 +1,72 @@
 import React, { useContext } from 'react'
-import { jsx } from '@emotion/core'
+import { jsx as emotion } from '@emotion/core'
 import styled from '@emotion/styled'
 import { ThemeProvider as EmotionProvider } from 'emotion-theming'
 import { MDXProvider } from '@mdx-js/react'
 import css from '@styled-system/css'
 import merge from 'lodash.merge'
 import get from 'lodash.get'
+import omit from 'lodash.omit'
 
 export { css } from '@styled-system/css'
 
+// custom pragma
+const spaceRegExp = /^[mp][trblxy]?$/
+const spaceProps = [
+  'm', 'mt', 'mr', 'mb', 'ml', 'mx', 'my',
+  'p', 'pt', 'pr', 'pb', 'pl', 'px', 'py',
+]
+
+const getSpace = props => {
+  if (!props) return undefined
+  const styles = {}
+  for (const key in props) {
+    if (!spaceRegExp.test(key)) continue
+    styles[key] = props[key]
+  }
+  return css(styles)
+}
+
+const hasSpace = props => {
+  for (const key in props) {
+    if (spaceRegExp.test(key)) return true
+  }
+  return false
+}
+
+const getCSS = props => {
+  if (!props) return undefined
+  if (!props.css && !hasSpace(props)) return undefined
+  return theme => [
+    typeof props.css === 'function'
+      ? css(props.css(theme))(theme)
+      : css(props.css)(theme),
+    getSpace(props)(theme),
+  ]
+}
+
+const system = (type, props, ...children) =>
+  React.createElement.apply(undefined, [
+    type,
+    {
+      ...omit(props, spaceProps),
+      css: getCSS(props)
+    },
+    ...children
+  ])
+
+// compose createElement functions
+const compose = (...funcs) => (type, props, ...children) => {
+  return funcs.reduce((el, fn) => {
+    return fn.apply(undefined, [ el.type, el.props, ...children ])
+  }, React.createElement.apply(undefined, [ type, props, ...children ]))
+}
+
+export const jsx = compose(
+  system,
+  emotion
+)
```

Comment:
FWIW I tried throwing the `mdx` create element function in here as well, and it *seemed* to work as expected, replacing e.g. `<h1>` with a component from context – something to think about as an alternative to the `Styled.h1` components

Reply:
That is really awesome. We should put some thought into using the mdx createElement function more generally (even though it seems to work).